src/dimension/Dimension.js

```diff
@@ -0,0 +1,488 @@
+    /******************************************************************
+     * private methods
+     */
+    /******************************************************************
+     * public methods
+     */
+    // todo - tricky: look at immerse/src/services/crossfilter.getTopN
+    group(groupId = false) { // todo - shouldn't this at least be parameterized like thus?
+        // if(!this._dimensionGroups.length) {
+        //     const newGroup = new Group(this.getDataConnector(), this)
+        //     return this.addGroupToDimension(newGroup)
+        // }
+        // else {
+        //     // todo - such an excellent question, eh? (tisws)
+        // }
+        const newGroup = new Group(this.getDataConnector(), this)
+        return this.addGroupToDimension(newGroup)
+    }
+    addGroupToDimension(newGroup) {
+        this._dimensionGroups.push(newGroup)
+        return newGroup
+    }
+    groupAll() {
+        return this.getCrossfilter().groupAll
+    }
+    setEliminateNull(eliminateNull) {
+        this._eliminateNull = eliminateNull
+        return this
+    }
+    getEliminateNull() {
+        return this._eliminateNull
+    }
+    /**
+     *  tbd public or private methods
+     */
+    multiDim (value) {
+        if (typeof value === "boolean") {
+            this._isMultiDim = value
+            return this
+        }
+        return this._isMultiDim
+    }
+    // todo - make this consistent with immerse/src/services/crossfilter.getTopN(),
+    // todo - which passes in 'column'
+    order(orderExpression) {
+        this._orderExpression = orderExpression
+        return this
+    }
+    orderNatural() {
+        this._orderExpression = null
+        return this
+    }
+    allowTargeted(allowTargeted) {
+        if (!arguments.length) {
+            return this._allowTargeted
+        }
+        this._allowTargeted = allowTargeted
+        return this // todo - returning 'this' is inconsistent
+    }
+    toggleTarget() {
+        const crossfilter = this.getCrossfilter()
+        if (crossfilter._targetFilter == this._dimensionIndex) { // TODO duplicates isTargeting
+            crossfilter._targetFilter = null // TODO duplicates removeTarget
+        } else {
+            crossfilter._targetFilter = this._dimensionIndex
+        }
+    }
+    removeTarget() {
+        const crossfilter = this.getCrossfilter()
+        if (crossfilter._targetFilter == this._dimensionIndex) {
+            crossfilter._targetFilter = null
+        }
+    }
+    isTargeting() {
+        const crossfilter = this.getCrossfilter()
+        return crossfilter._targetFilter === this._dimensionIndex
+    }
+    projectOn(expressions) {
+        this._projectExpressions = expressions
+        return this
+    }
+    projectOnAllDimensions(flag) {
+        this._projectOnAllDimensionsFlag = flag
+        return this
+    }
+    /** filter methods **/
+    selfFilter(_) { // todo - '_' is used to set _selfFilter, so it should have a semantic name!!!
```

Comment:
Agreed—what kind of an argument does `selfFilter` take? boolean or sql string would be my guess.